File: 20200215Python-China/0538. Convert BST to Greater Tree.py
'''
Given a Binary Search Tree (BST), convert it to a Greater Tree such that every key of the original BST is changed to the original key plus sum of all keys greater than the original key in BST.

Example:

Input: The root of a Binary Search Tree like this:
              5
            /   \
           2     13

Output: The root of a Greater Tree like this:
             18
            /   \
          20     13
Note: This question is the same as 1038: https://leetcode.com/problems/binary-search-tree-to-greater-sum-tree/

'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def convertBST(self, root: TreeNode) -> TreeNode:
        self.getConvertBST(root,[0])
        return root

    def getConvertBST(self,root,res):
        if not root:
            return
        if root.right:
            self.getConvertBST(root.right,res)
        res[0]+=root.val
        root.val=res[0]
        self.getConvertBST(root.left,res)

File: archives/C#/0538. Convert BST to Greater Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    int sum=0;
    public TreeNode ConvertBST(TreeNode root) {
        if (root==null){
            return root;
        }
        ConvertBST(root.right);
        root.val+=sum;
        sum=root.val;
        ConvertBST(root.left);
        return root;
    }
}
